### ISSUE TITLE
Support GCP Identity Aware Proxy in remote write configuration

### ISSUE THREAD

**Issue by TaylorMutch** [prometheus/prometheus#9240]
## Proposal
**Use case. Why is this important?**

For folks who run their own Prometheus remote write endpoints which are backed by GCP's HTTP(S) load balancers, being able to authenticate to a public endpoint using [Identity Aware Proxy with programmatic authentication](https://cloud.google.com/iap/docs/authentication-howto#authenticating_from_a_service_account) would be very convenient and easy for teams using GCP to adopt.

From what I can see in the remote config, the only necessary requirement would be to add a new `RoundTripper` that is able to obtain tokens using idtoken's [NewTokenSource](https://github.com/googleapis/google-api-go-client/blob/master/idtoken/idtoken.go#L67) and to add a new `iap` block in the `remote_config` section to configure the `audience` and potentially a GCP service account (optional, could be determined by default GCP credentials).

Would be interested if there are other considerations that should be made! I would also be interested in contributing this feature.

**Comment by roidelapluie** [comment]
Hello,

I am puzzled by this. I am not using this service at all, but it seems like this could be done by plain oauth2?

**Comment by roidelapluie** [comment]
By looking at the documentation, we should cover this already. I am not willing to add vendor specific ways to retrieve or read client ID/secret.

**Comment by TaylorMutch** [comment]
Perhaps I have misunderstanding of how to configure the oauth client for this use case - happy to use existing functionality, but I just haven't been able to find a working example for remote write on my own that uses GCP's identity aware proxy feature.

**Comment by TaylorMutch** [comment]
The issue I am hitting with this is being able to do something similar to what was implemented for the AWS [sigv4 signing](https://github.com/prometheus/prometheus/blob/main/storage/remote/sigv4.go#L58) in the remote client, which is vendor specific, so it seemed somewhat appropriate to be able to support GCP's method for authenticating GCP service accounts.

Happy to follow best guidance for how to do this right, but its unclear to me how this would be done otherwise.

**Comment by TaylorMutch** [comment]
So looked back at this again, and I was able to get this working in a different way. AFAIK It is not possible to do normal OAuth flow for service accounts due to how GCP's Identity Aware Proxy works with service accounts (which don't provide a client secret, but instead a PEM file) and requiring some custom claims. I'm not an expert on OAuth but there wasn't an obvious way for me to configure this in the Prometheus configuration.

The way I found to get around this was to provide a sidecar which can write a token file to a directory that Prometheus can read. I didn't find a sidecar application that did this, so I ended up [writing one myself](https://github.com/TaylorMutch/gcp-iap-token-writer). Then in the Prometheus configuration for remote write I just provide path to the access token:
```yaml
remote_write:
- url: https://thanos-receive.example.com/api/v1/receive
  headers:
    THANOS-TENANT: my-tenant
  authorization:
    credentials_file: /prometheus/auth/gcptoken
```

I will close this issue, hopefully others will find this solution helpful
